File: vect2.ts
'use strict'


class Vect2 {
    constructor(public x:number, public y:number) {
    }
    equals(obj) {
        return obj && this.x === obj.x && this.y === obj.y
    }
    hashCode() {
        return (this.x << 8) | this.y
    }
    
    sum(other) {
        if(isNaN(other))
            return this.sumi(other.x, other.y)
        else
            return this.sum(turns.toVector(other))
    }
    sumi(x, y) {
        return new Vect2(this.x + x, this.y + y)
    }
    
    distance4(other) {
        return Math.abs(this.x - other.x) + Math.abs(this.y - other.y)
    }
    
    rotated(angle) {
        var rad = turns.toRadians(angle)
        var sin = Math.sin(rad)
        var cos = Math.cos(rad)
        return new Vect2(this.x * cos - this.y * sin, this.x * sin + this.y * cos)
    }
    toString() {
        return this.x + ',' + this.y
    }
    
    static get ZERO() {
        return new Vect2(0, 0)
    }
    
    get inverse() {
        return new Vect2(-this.x, -this.y)
    }
}
module.exports = exports = Vect2

var turns = require('./turns') // must be after exports because of circular import

File: src/esquire/vect2.kt
package esquire

data class Vect2(val x: Double, val y: Double) {
    fun plus(other: Vect2) = Vect2(x + other.x, y + other.y)
    fun plusn(x: Double, y: Double) = Vect2(this.x + x, this.y + y)

    fun minus(other: Vect2) = Vect2(x - other.x, y - other.y)
    fun minus() = Vect2(-x, -y)

    fun times(other: Vect2) = Vect2(x * other.x, y * other.y)
    fun times(other: Double) = Vect2(x * other, y * other)

    fun div(other: Vect2) = Vect2(x / other.x, y / other.y)
    fun div(other: Double) = Vect2(x * other, y * other)

    fun dist4(other: Vect2) = Math.abs(x - other.x) + Math.abs(y - other.y)

    fun dist8(other: Vect2) = Math.max(Math.abs(x - other.x), Math.abs(y - other.y))

    fun rotated(angle: Double): Vect2 {
        val rad = turns.toRadians(angle)
        val sin = Math.sin(rad)
        val cos = Math.cos(rad)
        return Vect2(this.x * cos - this.y * sin, this.x * sin + this.y * cos)
    }
}

val ZERO = Vect2(0.0, 0.0)
